### DISCUSSION TITLE
searchParams Pagination prevents ISR / Static generation

### DISCUSSION THREAD

**Discussion by mkoreo** [vercel/next.js#58884] [App Router]
Designing my site with Next.js 13/14 App directory, I started of showing results with searchParams pagination.
ex. https://someblog.azerty/greatblogsposts?page=3

That is logical and clear.
However, it prevents any form of static generation for next.js because you have to use searchParams prop in a page.ts file.
The route turns into a dynamic route (which cannot use generateStaticParams as the searchParams page param cannot be adressed).

Should I opt to change it to https://someblog.azerty/greatblogsposts/page-3 instead for best SEO and performance?
Isn't that "weird"?

**Comment by moonkorea00** [comment]
@mkoreo have you managed to find a solution?

**Reply by mkoreo** [reply]
Hey,

I reworked my website to use the later format, without searchParams.
Ex. https://someblog.azerty/greatblogsposts/page/3

So no, I could not find a solution to have the searchParams statically generated.

**Reply by moonkorea00** [reply]
i think the later format is the solution you would want to use.

the [docs](https://nextjs.org/docs/app/api-reference/functions/use-search-params#static-rendering) suggests to use suspense with useSearchParams in a client component which keeps the route to be statically rendered while the dynamic part that uses useSearchParams is client-side rendered. But the fallback would be your initial HTML and the susepended component would render client side.

using searchParams would trigger dynamic rendering.

i found this [issue](https://github.com/vercel/next.js/discussions/58351) helpful as well.

**Reply by felipedeboni** [reply]
Yes, I was with an ugly workaround.

You can keep your querystring /whatever?page=x as long as you have a middleware that rewrites ?page=x to /whatever/page/x. If you ask me does not make any sense not being able to use search params with ISR, but it is what it is :)

**Comment by Ziothh** [comment]
Tbh this feels like a massive hole for websites that want to statically render for performance.

F.e.: I have a `/search` page that I want to statically render for good SEO. The content basically never changes.

I'd like to also statically render other variants of the page like `/search?category=5&page=2`.
These should just be computed once every day and stored in a CDN.

This isn't possible right now.

I know I can use `unstable_cache()` but there is still a performance difference.

**Comment by tlmader** [comment]
This is pretty negative for us—we use search params for some tab states. We had to move all of these out of search params to achieve ISR for landing pages. Why not at opt the base page without search params into ISR? Or better yet, to support the use cases above my comment, give us a way to be selective about which search param combinations (regex?) should be opted into ISR.

**Comment by Mohamed3on** [comment]
It's ridiculous that years later this is still an unsolved issue. You have to choose between either worse performance and unnecessary server work or ugly code workarounds (and URL designs).

**Comment by columk1** [comment]
This prevents prefetching too which kills performance on stores that have sorting/filtering/search. What's the recommended pattern for an ecommerce store to avoid this issue? The ecommerce template from Vercel is all dynamic so it's really slow to navigate.

I would expect a base path like `/products` with no params to be pre-rendered even if the page accesses `searchParams`.

**Reply by tlmader** [reply]
Pre-rendering the base path would solve our use cases for us as well.

**Reply by Kulimeya** [reply]
Yeah that would fix everything. I want the pagination to be pre-rendered on build so it can easily be crawled at a high rate at any time and if a user wants to sort the list, the searchParam gets added and I fetch the sorted data from my query wrapped in unstable_cache. 
But currently there seems to be no solution whatsoever to have pages pre-rendered on build and don't end up with completely ridiculous routes because you have to make every searchParam part of the route.
Please just let me use generateStaticParams even if a page uses searchParams and just generate the version without any. From there it's easy to work around.

**Reply by columk1** [reply]
I couldn't get `useSearchParams` to perform well so I ended up making the base path static by removing `searchParams` from `page.tsx` and then nesting a duplicate version of this page with `searchParams` under an arbitrary path segment. I added a function to `proxy.ts` to rewrite anything at `/products` that contains a query to `/products/query` under the hood. This way the first page of results doesn't suspend, but if a user does a search or visits the next page of results it will use the dynamic version.

I also combined a few of the suggestions from other comments to avoid suspending other pages. Such as replacing `?category=` with `/category/[slug]`.

I played around with making sort filters and pagination static but it was too messy.
